### ISSUE TITLE
Support for Lumen framework

### ISSUE THREAD

**Issue by nunomaduro** [nunomaduro/larastan#26]
Would be good add support to Lumen.

https://medium.com/@rafaelcanical/although-i-agree-with-jan-alfred-richter-this-is-cool-can-this-be-used-with-lumen-projects-also-259b64c3d10c

**Comment by goodevilgenius** [comment]
I noticed that it states that it depends on `laravel/framework`. This is probably overkill, and likely the biggest hurdle to Lumen compatibility.

I imagine changing that to `illuminate/console` (and perhaps a few other needed illuminate packages) would be sufficient, and would allow compatibility with both Laravel and Lumen.

**Comment by nunomaduro** [comment]
@goodevilgenius I agree with you. Do you mind of prepare a pull request for that? Please take in consideration all the components/classes of Laravel that are already used on the package.

**Comment by goodevilgenius** [comment]
I don't have the time to go through the `larastan` code and find which `illuminate` packages are specifically needed, sadly.

**Comment by nunomaduro** [comment]
**How to resolve this issue:**

1. Identity the components illuminate used on the package.
2. Include Laravel components on the composer.json instead of the full framework.
3. Test with Lumen locally.

Example before:

```
    "require": {
        "php": "^7.1.3",
        "phpstan/phpstan": "^0.10",
        "symfony/process": "^4.1",
        "laravel/framework": "5.6.*",
        "mockery/mockery": "^1.0|0.9.*"
    },
```

Example after:

```
    "require": {
        "php": "^7.1.3",
        "phpstan/phpstan": "^0.10",
        "symfony/process": "^4.1",
        "illuminate/support": "5.6.*",
        "illuminate/database": "5.6.*",
        "mockery/mockery": "^1.0|0.9.*"
    },
```

**Comment by nunomaduro** [comment]
@richardkeep  The version `v0.2.0` should support Lumen and Laravel Zero. Do you mind of running some tests? I don't use Lumen that much.

**Comment by hskrasek** [comment]
@nunomaduro We use Lumen at work, I can give it a try this evening and see how it works out.

**Comment by hartman** [comment]
Working pretty nicely, but I have errors on Mixins.php line 71
> Class Eloquent does not exist 

Which is probably related to some of the laravel ide-helper magic...

**Comment by nunomaduro** [comment]
@hartman Ok! I will update the readme.

@hskrasek You got news?

**Comment by hskrasek** [comment]
@nunomaduro Yesterday turned into a meeting and weird production bug day so I haven’t had a chance to play with it. I’m hoping to today

**Comment by hskrasek** [comment]
So our project has some.. unique constraints due to legacy code dependencies. Because of this I can’t get s good test without doing some overriding of how phpstan handles composer autoloading.

If I can hack around that I’ll be able to give you a better test. And probably a feature request lol

**Comment by goodevilgenius** [comment]
My project is a pretty standard Lumen project. I'll give it a try tomorrow when I get in to work.

**Comment by hartman** [comment]
@goodevilgenius that sounds like #40

I'm still looking for hints as to why it crashes on my usage of the Eloquent facade, if anyone has any ideas, i'd love to hear them.

**Comment by luispabon** [comment]
absolutely latest as installed 5 minutes ago

**Comment by luispabon** [comment]
v0.2.4

**Comment by luispabon** [comment]
The interface is implemented by several models, same as the other one that doesn't make it crash.

**Comment by hartman** [comment]
> I'm still looking for hints as to why it crashes on my usage of the Eloquent facade, if anyone has any ideas, i'd love to hear them.

So this is the Eloquent mixin phpdoc annotation that is added by @barryvdh/laravel-ide-helper 

It cannot resolve the mixin, because basically it doesn't know it exists. I'm not even entirely sure if this error is now within ide-helper, or within the class lookup of larastan actually. It doesn't seem like this is easily ignorable, but i added a try catch to getMixinsForClass to ignore the reflection class error. I couldn't figure out a cleaner way to handle this problem.

```
$mixins = array_filter(
            $mixins,
            function ($mixin) use ($classReflection) {
                try {
                    return (new \ReflectionClass($mixin))->getName() !== $classReflection->getName();
                } catch (\ReflectionException $e) {
                }
            }
        );
```
Is there a cleaner way ?

**Comment by Korri** [comment]
@nunomaduro Looks like #155 also has the secondary effect of installing laravel/framework on Lumen projects.

**Comment by nunomaduro** [comment]
@Korri There is a problem on that? This is a dev dependency.

**Comment by Korri** [comment]
@nunomaduro The only issue here is `orchestra/testbench` if there is a way to not make it an hard dependency (only require in when testing a package), that would solve the issue.

**Comment by nunomaduro** [comment]
@Korri I may have an idea, tomorrow I will code it 👍

**Comment by nunomaduro** [comment]
No solution so far.

**Comment by Korri** [comment]
Looks like the only hard requirement is the `Orchestra\Testbench\Concerns\CreatesApplication` trait, is there a way to replace it with something that is compatible with Lumen and Laravel?

**Comment by ian-nisbet** [comment]
It works fine with Lumen already.

I only had to add in the Service Provider to bootstrap/app.php
`$app->register(NunoMaduro\Larastan\LarastanServiceProvider::class);`

and specify the path.storage and path.config
`$app->instance('path.storage', app()->basePath() . DIRECTORY_SEPARATOR . 'storage');`
// https://github.com/nasrulhazim/lumen-in-laravel/issues/1
`$app->instance('path.config', app()->basePath() . DIRECTORY_SEPARATOR . 'config');`
// https://stackoverflow.com/questions/38961790/class-path-config-does-not-exist-lumen-5-2

**Comment by vojtasvoboda** [comment]
@ian-nisbet Could you please write down step-by-step tutorial how to do that with Lumen? Thank you

I'm still getting `orchestra/testbench 3.6.x-dev requires laravel/framework ~5.6.34`.

**Comment by sksethi25** [comment]
@nunomaduro Hey, Any update on this guys?

**Comment by nunomaduro** [comment]
@sksethi25 I haven't work on this. Fell free to propose a solution/pr.

**Comment by sksethi25** [comment]
@nunomaduro  as of now i just forked and moved "orchestra/testbench" into dev dependence and installed forked version with latest Lumen (5.7.7) (Laravel Components 5.7.*) . Registered service provider and ran the command, it worked like a charm.I have now alot of things to fix first,But Two of them of which caught any immediate attention are :
 

> Call to an undefined method Illuminate\Database\Query\Builder::withTrashed()

and

>  Internal error: Call to a member function connection() on null  
>          Run PHPStan with --debug option and post the stack trace to:    
>          https://github.com/phpstan/phpstan/issues/new     

if you have any ideas about these , let me know. i will too try to update on whats causing these.

**Comment by nunomaduro** [comment]
For the `Call to an undefined method Illuminate\Database\Query\Builder::withTrashed()` is a duplicated issue, that already exists.

For the second you need to sahre the code where that happens.

**Comment by TheLevti** [comment]
Seems to break the whole app if you do not install it (production) by using `composer install --no-dev`.

Whole app fails when reaching the following lines in `bootstrap/app.php`
```
$app = new Laravel\Lumen\Application(
    realpath(__DIR__.'/../')
);
```

**Comment by TheLevti** [comment]
All I get is just an empty 200 response for any endpoint. Also php artisan then just does nothing.

**Comment by j3rrey** [comment]
Working perfectly fine on my end.
Lumen 5.8 
Fairly big project for a lumen app

**Comment by TheLevti** [comment]
Does your app work if you install with ---no-dev?

**Comment by j3rrey** [comment]
> Does your app work if you install with ---no-dev?

Well I have a production switch, so I assume yes.

**Comment by TheLevti** [comment]
Thought so the same, but apparently it does not work for me. I also have it in the dev requirements. Just wanted to ask if someone else tried it.

**Comment by j3rrey** [comment]
@TheLevti 
hmm in that case, how are you registering the service provider?

**Comment by j3rrey** [comment]
@TheLevti try to do a ```composer dump-auto``` after your initial composer install --no-dev

**Comment by sudomabider** [comment]
Surprised no one else has mentioned @vojtasvoboda 's problem yet.

```
    "require": {
        "php": "^7.1.3",
        "symfony/process": "^4.2",
        "phpstan/phpstan": "^0.11.8",
        "orchestra/testbench": "^3.6",
```

`orchestra/testbench` requires `laravel/framework` and is causing problems with installing in Lumen. Why is `orchestra/testbench` required at runtime anyway? As far as I can see it's only used in tests (plus an unused import in `ApplicationResolver.php` which probably shouldn't be there). Moving `orchestra/testbench` to `require-dev` should allow us to proceed further with Lumen. Or am I missing something here?

**Comment by goodevilgenius** [comment]
@sudomabider I noticed. I removed `larastan` from my project on our last upgrade because of that.

**Comment by nunomaduro** [comment]
https://github.com/nunomaduro/larastan/issues/295

**Comment by nunomaduro** [comment]
It's fixed on the `master` branch. The fix will be deployed on the `4.0` version.
